### DISCUSSION TITLE
Welcome to PyNFe Discussions!

### DISCUSSION THREAD

**Discussion by leotada** [TadaSoftware/PyNFe#95] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by nilton-medeiros** [comment]
Olá, pessoal, estou trabalhando nessa biblioteca PyNFe para emissão de NFe, NFCe, CTe e MDFe em modo produção, espero num futuro próximo compreendê-la e poder colaborar com experiências reais com a PyNFe. Trabalho com CTe desde 2013.
Na linha 135 do processamento/serializers.py na "def _serializar_cliente", é atribuído ao "tipo_documento" o próprio número do documento ao invés do tipo como 'CNPJ', 'CPF' ou 'idEstrangeiro' (caso idDest = 3). Alguma explicação para isso? Não consegui achar a lógica varrendo as fontes. Isso também ocorre na "_serializar_transportadora".

**Reply by leogregianin** [reply]
https://github.com/TadaSoftware/PyNFe/blob/5cfc800ee2080dafb6857203b0c7cde7e22fa00a/pynfe/processamento/serializacao.py#L135

@nilton-medeiros, no ```cliente.tipo_documento``` você define o tipo conforme o layout exemplo: 'CNPJ', 'CPF' ou 'idEstrangeiro' e no ```cliente.numero_documento``` você define o número do CPF, o número do CNPJ. E da mesma forma ocorre no grupo de transportadora, entrega e retirada.

**Reply by nilton-medeiros** [reply]
Exato, é essa a ideia e a lógica, mas o que eu vejo é que o cnpj está sendo atribuido ao tipo de documento, veja:
```etree.SubElement(raiz, cliente.tipo_documento).text = so_numeros(cliente.numero_documento)```
cliente.tipo_documento recebe os números do documento, isso não é uma comparação e sim uma atribuição. Isso não é um erro? Se for isso erro ocorre no emitente e na transportadora.

**Reply by leogregianin** [reply]
Veja um exemplo:

```python
from pynfe.entidades.cliente import Cliente
(...)


variavel_cpf_cnpj = '00011141400'

if len(variavel_cpf_cnpj) == 11:
    variavel_tipo_documento = 'CPF'
else:
    variavel_tipo_documento = 'CNPJ'

cliente = Cliente(
    razao_social="",
    email="",
    indicador_ie="",
    tipo_documento=variavel_tipo_documento,
    numero_documento=variavel_cpf_cnpj
    (...)   
```

**Reply by nilton-medeiros** [reply]
Não amigo, vc não entendeu o que estou tentando dizer. Não estou com dúvida como atribuir os valores corretos os atributos da classe.
Estou informando que há algo errado no código fonte na linha 135 do processamento/serializers.py na "def _serializar_cliente" onde deveria atribuir o tipo do documento (CNPJ/CPF..) mas está atribuindo o próprio número do CNPJ/CPF...

**Reply by nilton-medeiros** [reply]
Esta é a linha do código fonte da PyNFe em processamento/serializers.py, NÃO é o meu código:
```etree.SubElement(raiz, cliente.tipo_documento).text = so_numeros(cliente.numero_documento)```

**Reply by juniortada** [reply]
@nilton-medeiros você está fazendo uma pequena confusão entre NOME da tag do xml e CONTEÚDO (valor) preenchido
na tag.
Esta linha específica que você está citando, após o processamento da serialização ficaria assim:

```python
 # <tipo_documento> --- # valor do documento </tipo_documento>
<CPF>123456789</CPF>
 ou ainda
<CNPJ>1231231230001-00</CNPJ>
```

a atribuição que você se refere é ao número (valor) do documento.
cliente.tipo_documento é a NOMEAÇÃO (definição do texto que da NOME na tag)

**Reply by nilton-medeiros** [reply]
Verdade @juniortada!! Tem razão, agora entendi, desculpem minha confusão, nesse passo refere-se as tags do xml... obrigado pelo esclarecimento.

**Comment by nilton-medeiros** [comment]
Olá, rapazes, no arquivo notafiscal.py em pynfe/entidades, linha 77 o atributo "forma_emissao" está como string, mas deveria ser inteiro conforme NF_FORMAS_EMISSAO, certo?
```
# - Forma de emissao (obrigatorio - seleciona de lista) - NF_FORMAS_EMISSAO
forma_emissao = str()

NF_FORMAS_EMISSAO = (
    (1, 'Normal'),
    (2, 'Contingencia'),
    (3, 'Contingencia com SCAN'),
    (4, 'Contingencia via DPEC'),
    (5, 'Contingencia FS-DA'),
)
```

**Reply by leogregianin** [reply]
A propriedade ```forma_emissao``` receberá o ```str``` diretamente, você pode conferir no [wiki](https://github.com/TadaSoftware/PyNFe/wiki/Emiss%C3%A3o-de-NFe) como atribuir cada campo.

Veja que nos ```imports``` do ```notafiscal.py``` não utiliza este enum NF_FORMAS_EMISSAO. Poderia usar, mas atualmente não utiliza.

**Reply by nilton-medeiros** [reply]
@leogregianin, muito obrigado pelo esclarecimento.

**Comment by dslima90** [comment]
Amigos, aglumas dúvidas, comecei agora a tentar entender como funciona o serviço de emissão de nfe.

1 Como consigo um certificado para testes? 

2Em produção, preciso de um certificado para cada CNPJ ou um certificado me permite emitir nfe para multiplos cnpjs?

**Reply by juniortada** [reply]
Boa tarde @dslima90 
1 - Não existe certificado válido para testes em ambiente de homologação, apenas certificado "real".
2 - O certificado digital é único para cada CNPJ, seja para produção ou homologação. Cada emissor de Nota 
precisa do seu respectivo certificado.

**Reply by dslima90** [reply]
Obrigado pelas respostas!

**Comment by developerAdriano** [comment]
Alguém também teve dúvidas sobre os campos sem orientação ?

**Comment by valdecirsanches** [comment]
Olá amigos eu tambem embarquei nesta. peguei a biblioteca para utilizar em meu projeto e já estou animadissimo!!

Alguem ja fez autorização para SP ????
Estou tendo problemas:
Rejeição: Falha no Schema XML do lote de NFe

notei que no xml não esta a tag "cobr"

![image](https://user-images.githubusercontent.com/17271010/224124340-c55e3923-a475-4f29-8f6a-d678c54e4021.png)
![image](https://user-images.githubusercontent.com/17271010/224124584-3960db3a-07b3-4ef1-ba84-a119a5379500.png)

**Reply by valdecirsanches** [reply]
Resolvido!!!  Agora outro  erro que não consigo resolver: 

![image](https://user-images.githubusercontent.com/17271010/224194146-7f6e1ec4-6742-4620-9b50-aacc284c35f1.png)

![image](https://user-images.githubusercontent.com/17271010/224194438-5c74b8e4-76a7-4567-a727-b40cd3bcd4a2.png)

**Reply by valdecirsanches** [reply]
Resolvido era uma falha minha... desculpem-me
Conferi o xml e achei o problema
